generator/default.p9k

```diff
@@ -0,0 +1,606 @@
+#   $2 string State of the segment.
+#   $3 string Alignment (left|right).
+#   $4 integer Index of the segment.
+#   $5 bool Whether the segment should be joined.
+#   $6 string Background color.
+#   $7 string Foreground color.
+#   $8 string Content of the segment.
+#   $9 string Visual identifier (must be a key of the icons array).
+#   $10 string The condition - if the segment should be shown (gets evaluated).
+##
+serializeSegment() {
+  local NAME="${1}"
+  local STATE="${2}"
+  local ALIGNMENT="${3}"
+  local INDEX="${4}"
+  local JOINED="${5}"
+
+  local STATEFUL_NAME="${(U)NAME#prompt_}"
+  [[ -n "${STATE}" ]] && STATEFUL_NAME="${STATEFUL_NAME}_${(U)STATE}"
+
+  # Overwrite given background-color by user defined variable for this segment.
+  local BACKGROUND_USER_VARIABLE="P9K_${STATEFUL_NAME}_BACKGROUND"
+  local BACKGROUND="${(P)BACKGROUND_USER_VARIABLE}"
+  [[ -z "${BACKGROUND}" ]] && BACKGROUND="${6}"
+
+  # Overwrite given foreground-color by user defined variable for this segment.
+  local FOREGROUND_USER_VARIABLE="P9K_${STATEFUL_NAME}_FOREGROUND"
+  local FOREGROUND="${(P)FOREGROUND_USER_VARIABLE}"
+  [[ -z "${FOREGROUND}" ]] && FOREGROUND="${7}"
+
+  # Overwrite given bold directive by user defined variable for this segment.
+  local BOLD_USER_VARIABLE="P9K_${STATEFUL_NAME}_BOLD"
+  local BOLD="${(P)BOLD_USER_VARIABLE}"
+  [[ -z "${BOLD}" ]] && BOLD=false || BOLD=true
+
+  local CONTENT="${8}"
+
+  local VISUAL_IDENTIFIER
+  if [[ -n "${9}" ]]; then
+    VISUAL_IDENTIFIER="$(printIcon ${9})"
+    if [[ -n "${VISUAL_IDENTIFIER}" ]]; then
+      # Allow users to overwrite the color for the visual identifier only.
+      local visual_identifier_color_variable="P9K_${STATEFUL_NAME}_VISUAL_IDENTIFIER_COLOR"
+      setDefault "${visual_identifier_color_variable}" "${FOREGROUND}"
+      VISUAL_IDENTIFIER="%F{${(P)visual_identifier_color_variable}%}${VISUAL_IDENTIFIER}%f"
+      # Add an whitespace if we print more than just the visual identifier
+      if [[ -n "${CONTENT}" ]]; then
+        [[ "${ALIGNMENT}" == "left" ]] && VISUAL_IDENTIFIER="${VISUAL_IDENTIFIER} "
+        [[ "${ALIGNMENT}" == "right" ]] && VISUAL_IDENTIFIER=" ${VISUAL_IDENTIFIER}"
+      fi
+    fi
+  fi
+  # Conditions have three layers:
+  # 1. All segments should not print
+  #    a segment, if they provide no
+  #    content (default condition).
+  # 2. All segments could define a
+  #    default condition on their
+  #    own, overriding the previous
+  #    one.
+  # 3. Users could set a condition
+  #    for each segment. This is
+  #    the trump card, and has
+  #    highest precedence.
+  local CONDITION
+  local SEGMENT_CONDITION="P9K_${STATEFUL_NAME}_CONDITION"
+  if defined "${SEGMENT_CONDITION}"; then
+    CONDITION="${(P)SEGMENT_CONDITION}"
+  elif [[ -n "${10}" ]]; then
+    CONDITION="${10}"
+  else
+    CONDITION='[[ -n "${8}" ]]'
+  fi
+  # Precompile condition.
+  eval "${CONDITION}" && CONDITION=true || CONDITION=false
```

Comment:
This whole condition block is taken from my [async_all_the_segments](https://github.com/dritter/powerlevel9k/blob/async_all_the_segments/powerlevel9k.zsh-theme#L1458-L1480)-Branch, right?
So with this code it is possible to define own conditions for each segment. That my be worth noticing in the changelog and readme.

Reply:
Yes, I used your branch as a base long ago, so that is a carry-over. Would you like to write a description for that?

Reply:
Not sure if we misunderstood us here: I wanted a description for the condition-overriding mechanism.
Something like:
> ### Defining own conditions
> By default segments are only rendered if they have content. If you want to change that, you could define a variable that follows the naming scheme `P9K_${SEGMENT_NAME}_CONDITION` (where `${SEGMENT_NAME}` is the uppercase name of the segment you want to put the condition on). If you have a [stateful segment](https://github.com/bhilburn/powerlevel9k/wiki/Stylizing-Your-Prompt#special-segment-colors), the naming scheme changes as follows: `P9K_${SEGMENT_NAME}_${STATE}_CONDITION`.

Btw. We should rename the wiki section as well (the core of that is to explain states).